### PR TITLE
Fix token choice w/ checkpointing, add tests

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
Now token choice remembers its discrete choices.

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
           Kinda, adding switch tests could be split, but IMO it's a nice extra
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] ~I attached link to neptune if it was necessary~
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

**Comment by Simontwice** [comment]
The tests pass, but does it run? I'd like a neptune run for 10 steps

File: research/conditional/moe_layers/token_choice.py [hunk 1]
```diff
@@ -90,11 +92,28 @@ def forward(self, x: torch.Tensor):
...
+                    with torch.no_grad():
+                        expert_index = self._checkpointed_expert_index
+                        assert isinstance(expert_index, torch.Tensor)
+
+                expert_gate = torch.gather(
+                    gate_out, dim=1, index=expert_index
+                ).squeeze()
+                expert_index = expert_index.squeeze()
+            else:
+                expert_gate, expert_index = torch.max(gate_out, dim=1)
```
Comment (crewtool) [inline research/conditional/moe_layers/token_choice.py:116]: Wrzućmy to do oddzielnej funkcji.
```
def choose_expert(gate_out):
   checkpointing_enabled = (
            checkpointing.is_in_first_forward() or checkpointing.is_in_second_forward()
        )
  if checkpointing_enabled:
                if checkpointing.is_in_first_forward():
                    with torch.no_grad():
                        expert_index = torch.argmax(gate_out, dim=1, keepdim=True)
                        self._checkpointed_expert_index = expert_index

                if checkpointing.is_in_second_forward():
                    with torch.no_grad():
                        expert_index = self._checkpointed_expert_index
                        assert isinstance(expert_index, torch.Tensor)

                expert_gate = torch.gather(
                    gate_out, dim=1, index=expert_index
                ).squeeze()
                expert_index = expert_index.squeeze()
            else:
                expert_gate, expert_index = torch.max(gate_out, dim=1)
  return expert_gate, expert_index
```

File: research/conditional/moe_layers/token_choice.py [hunk 2]
```diff
@@ -184,6 +183,27 @@ def forward(self, x: torch.Tensor):
...
 
+    def choose_expert(self, gate_out) -> tuple[torch.Tensor, torch.Tensor]:
+        checkpointing_enabled = (
+            checkpointing.is_in_first_forward() or checkpointing.is_in_second_forward()
+        )
+        if checkpointing_enabled:
+            if checkpointing.is_in_first_forward():
+                with torch.no_grad():
+                    expert_index = torch.argmax(gate_out, dim=1, keepdim=True)
+                    self._checkpointed_expert_index = expert_index
```
Comment (Simontwice) [inline research/conditional/moe_layers/token_choice.py:194]: ape-brain question: why do we use argmax if checkpointing else max?

Comment (maciek-pioro) [reply research/conditional/moe_layers/token_choice.py:194]: max and argmax return different thins

Comment (Simontwice) [reply research/conditional/moe_layers/token_choice.py:194]: I realise this. Was hoping for a brief overview of how the approach differs in the two cases. Read it myself in the end.

**Review by Simontwice — APPROVED** [review]
ok

**Comment by maciek-pioro** [comment]
Merged in another PR

**Pull request closed without merging**